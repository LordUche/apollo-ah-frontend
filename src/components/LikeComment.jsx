import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { postLikeComment, postDislikeComment } from '../actions/likeCommentsAction';
// import debounceFn from '../utils/debounce';

class LikeComment extends React.Component {
  static propTypes = {
    postLikeComment: PropTypes.func.isRequired,
    postDislikeComment: PropTypes.func.isRequired,
  }

  state = {
    likeCount: 0,
    dislikeCount: 0,
    id: '',
    slug: '',
    clickedLike: false,
    clickedDislike: false,
  }


  static getDerivedStateFromProps(nextProps, prevState) {
    const {
      likeCount, dislikeCount, id, slug,
    } = nextProps;
    if (likeCount < prevState.likeCount) return null;
    if (dislikeCount < prevState.dislikeCount) return null;
    return {
      likeCount, dislikeCount, id, slug,
    };
  }

  handleLikes = () => {
    const { id, slug, clickedLike } = this.state;
    if (!clickedLike) this.setState(prevState => ({ likeCount: prevState.likeCount + 1 }));
    if (clickedLike) this.setState(prevState => ({ likeCount: prevState.likeCount - 1 }));
    const { postLikeComment: likeComment } = this.props;
    this.setState(prevState => ({ clickedLike: !prevState.clickedLike }));
    likeComment({ id, slug });
  }

  handleDislikes = () => {
    const { clickedDislike, id, slug } = this.state;
    if (!clickedDislike) this.setState(prevState => ({ dislikeCount: prevState.dislikeCount + 1 }));
    if (clickedDislike) this.setState(prevState => ({ dislikeCount: prevState.dislikeCount - 1 }));
    const { postDislikeComment: dislikeComment } = this.props;
    this.setState(prevState => ({ clickedDislike: !prevState.clickedDislike }));
    dislikeComment({ id, slug });
  }

  render() {
    const { likeCount, dislikeCount } = this.state;
    return (
      <div className="like-comment-parent">
        <div>
          <i
            className="fas fa-thumbs-up"
            onClick={this.handleLikes}
            role="button"
            aria-hidden
          />
          {<span>{likeCount}</span>}
        </div>
        <div>
          <i
            className="fas fa-thumbs-down"
            onClick={this.handleDislikes}
            role="button"
            aria-hidden
          />
          {<span>{dislikeCount}</span>}
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = { postDislikeComment, postLikeComment };

export default connect(() => { }, mapDispatchToProps)(LikeComment);
// export default LikeComment;
