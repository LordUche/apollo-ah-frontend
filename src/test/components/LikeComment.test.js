import React from 'react';
import setup from '../setup';
import LikeComment from '../../components/LikeComment';

const mockFn = jest.fn();
const _props = {
  id: 2,
  slug: 'djsjhdskl-sdmnds',
  likeCount: 3,
  dislikeCount: 9,
  postDislikeComment: mockFn,
  postLikeComment: mockFn
};

describe('<LikeComment />', () => {
  it('Should render the component', () => {
    const wrapper = setup(
      <LikeComment
        {..._props}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('Should render the component', () => {
    const wrapper = setup(<LikeComment {..._props} />);
    const btn = wrapper.find('.fa-thumbs-down');
    const btn2 = wrapper.find('.fa-thumbs-up');
    btn.simulate('click', { handleDislikes: f => f });
    btn2.simulate('click', { handleLikes: f => f });
    expect(btn).toBeDefined();
    expect(btn2).toBeDefined();
  });
  it('State should change on button click', () => {
    const wrapper = setup(<LikeComment {..._props} />);
    const LikeCommentWrapper = wrapper.find('LikeComment');
    LikeCommentWrapper.setState({ clickedLike: false });
    LikeCommentWrapper.setState({ clickedDislike: false });
    const btn = wrapper.find('.fa-thumbs-down');
    const btn2 = wrapper.find('.fa-thumbs-up');
    btn.simulate('click');
    btn2.simulate('click');
    expect(LikeCommentWrapper.state().clickedDislike).toBe(true);
    expect(LikeCommentWrapper.state().clickedLike).toBe(true);
    btn.simulate('click');
    btn2.simulate('click');
    expect(LikeCommentWrapper.state().clickedDislike).toBe(false);
    expect(LikeCommentWrapper.state().clickedLike).toBe(false);
  });
});
