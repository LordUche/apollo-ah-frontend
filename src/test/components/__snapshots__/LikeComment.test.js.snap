// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LikeComment /> Should render the component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(LikeComment)
        dislikeCount={9}
        id={2}
        likeCount={3}
        postDislikeComment={[MockFunction]}
        postLikeComment={[MockFunction]}
        slug="djsjhdskl-sdmnds"
      >
        <LikeComment
          dislikeCount={9}
          id={2}
          likeCount={3}
          postDislikeComment={[Function]}
          postLikeComment={[Function]}
          slug="djsjhdskl-sdmnds"
        >
          <div
            className="like-comment-parent"
          >
            <div>
              <i
                aria-hidden={true}
                className="fas fa-thumbs-up"
                onClick={[Function]}
                role="button"
              />
              <span>
                3
              </span>
            </div>
            <div>
              <i
                aria-hidden={true}
                className="fas fa-thumbs-down"
                onClick={[Function]}
                role="button"
              />
              <span>
                9
              </span>
            </div>
          </div>
        </LikeComment>
      </Connect(LikeComment)>
    </Router>
  </BrowserRouter>
</Provider>
`;
